source/common/router/router.cc

```diff
@@ -391,6 +417,25 @@ Http::FilterHeadersStatus Filter::decodeHeaders(Http::HeaderMap& headers, bool e
     sendNoHealthyUpstreamResponse();
     return Http::FilterHeadersStatus::StopIteration;
   }
+  if (debug_config && debug_config->append_upstream_host_) {
+    // The hostname and address will be appended to any local or upstream responses from this point,
+    // possibly in addition to the cluster name.
+    modify_headers = [modify_headers, debug_config, conn_pool](Http::HeaderMap& headers) {
+      modify_headers(headers);
+      headers.addCopy(Http::LowerCaseString(
+                          debug_config->hostname_header_.value_or("x-envoy-upstream-hostname")),
```

Comment:
Perhaps add these header constants to headers.h?

Reply:
Done.